remote.go

```diff
@@ -0,0 +1,265 @@
+
+		if _, err := body.Seek(0, io.SeekStart); err != nil {
+			return fmt.Errorf("body seek failed: %v", err)
+		}
+
+		if perr != nil {
+			rd.Log.Debugf("deliverForServer: %+v (delivery ID = %s)", perr, ctx.DeliveryID)
+			for _, successful := range perr.Successful {
+				partialErr.Successful = append(partialErr.Successful, successful)
+			}
+			for _, temporaryFail := range perr.TemporaryFailed {
+				partialErr.TemporaryFailed = append(partialErr.TemporaryFailed, temporaryFail)
+			}
+			for _, failed := range perr.Failed {
+				partialErr.Failed = append(partialErr.Failed, failed)
+			}
+			for k, v := range perr.Errs {
+				partialErr.Errs[k] = v
+			}
+		}
+
+	}
+
+	if len(partialErr.Failed) == 0 && len(partialErr.TemporaryFailed) == 0 {
+		return nil
+	}
+	return partialErr
+}
+
+func toPartialError(temporary bool, rcpts []string, err error) *PartialError {
+	perr := PartialError{
+		Errs: make(map[string]error, len(rcpts)),
+	}
+	if temporary {
+		perr.TemporaryFailed = rcpts
+	} else {
+		perr.Failed = rcpts
+	}
+	for _, rcpt := range rcpts {
+		perr.Errs[rcpt] = err
+	}
+	return &perr
+}
+
+func (rd *RemoteDelivery) deliverForServer(ctx *module.DeliveryContext, domain string, rcpts []string, body io.Reader) *PartialError {
+	addrs, err := lookupTargetServers(domain)
+	if err != nil {
+		return toPartialError(false, rcpts, err)
+	}
+
+	var cl *smtp.Client
+	var lastErr error
+	var usedServer string
+	for _, addr := range addrs {
+		cl, err = connectToServer(rd.hostname, addr, rd.requireTLS)
+		if err != nil {
+			rd.Log.Debugf("failed to connect to %s: %v", addr, err)
+			lastErr = err
+			if !isTemporaryErr(err) {
+				break
+			}
+			continue
+		}
+		usedServer = addr
+	}
+	if cl == nil {
+		return toPartialError(isTemporaryErr(lastErr), rcpts, lastErr)
+	}
+
+	rd.Log.Debugln("connected to", usedServer)
+
+	if err := cl.Mail(ctx.From); err != nil {
+		rd.Log.Printf("MAIL FROM failed: %v (server = %s, delivery ID = %s)", err, usedServer, ctx.DeliveryID)
+		return toPartialError(isTemporaryErr(err), rcpts, err)
+	}
+
+	perr := PartialError{Errs: make(map[string]error)}
+	for _, rcpt := range rcpts {
+		if err := cl.Rcpt(rcpt); err != nil {
+			rd.Log.Printf("RCPT TO failed: %v (server = %s, delivery ID = %s)", err, usedServer, ctx.DeliveryID)
+			if isTemporaryErr(err) {
+				perr.TemporaryFailed = append(perr.TemporaryFailed, rcpt)
+			} else {
+				perr.Failed = append(perr.Failed, rcpt)
+			}
+			perr.Errs[rcpt] = err
+		}
+	}
+	bodyWriter, err := cl.Data()
+	if err != nil {
+		rd.Log.Printf("DATA failed: %v (server = %s, delivery ID = %s)", err, usedServer, ctx.DeliveryID)
+		return toPartialError(isTemporaryErr(err), rcpts, err)
+	}
+	defer bodyWriter.Close()
+	if _, err := io.Copy(bodyWriter, body); err != nil {
+		rd.Log.Printf("body write failed: %v (server = %s, delivery ID = %s)", err, usedServer, ctx.DeliveryID)
+		return toPartialError(isTemporaryErr(err), rcpts, err)
+	}
+
```

Comment:
Can we close the `bodyWriter` here and check the error? (It should be fine to double-close and not check the second close error in `defer`)

Reply:
Done.